parser/charset/encoding.go

```diff
@@ -140,9 +141,40 @@ func (e *Encoding) EncodeInternal(dest, src []byte) []byte {
 	return dest
 }
 
+// validUTF8 checks whether there are illegal utf8 characters in []byte.
+func (e *Encoding) validUTF8(src []byte) ([]byte, error) {
+	resultBytes := src
+	for len(src) > 0 {
+		r, size := utf8.DecodeRune(src)
+		if r == utf8.RuneError && size == 1 {
+			return resultBytes, e.generateErr(src, 3)
+		}
+		src = src[size:]
+	}
+
+	return resultBytes, nil
+}
+
+// validUTF8String checks whether there are illegal utf8 characters in string.
+func (e *Encoding) validUTF8String(src string) (string, error) {
+	resultStr := src
+	for len(src) > 0 {
+		r, size := utf8.DecodeRuneInString(src)
+		if r == utf8.RuneError && size == 1 {
+			return resultStr, e.generateErr([]byte(src), 3)
+		}
+		src = src[size:]
+	}
+
+	return resultStr, nil
+}
+
 // Decode convert bytes from a specific charset to utf-8 charset.
 func (e *Encoding) Decode(dest, src []byte) ([]byte, error) {
 	if !e.enabled() {
+		if e.name == "utf-8" {
```

Comment:
Use `encodingLegacy` instead of `utf-8`?

Reply:
I think `utf-8` is better. `encodingLegacy` may change someday, and `validUTF8` is only design for `utf8`.